app/models/search/index_sweeper.rb

```diff
@@ -0,0 +1,99 @@
+class IndexSweeper
+
+  REDIS = AsyncIndexer::REDIS
+
+  def self.async_cleanup(klass, expected_ids, found_ids)
+    deleted_ids = expected_ids.map(&:to_i).select { |id| !found_ids.include?(id) }
+
+    if deleted_ids.any?
+      AsyncIndexer.index(klass, deleted_ids, "cleanup")
+    end
+  end
+
+  def initialize(batch, indexer)
+    @batch = batch
+    @indexer = indexer
+    @rerun_ids = []
+
+    ensure_failure_stores
+  end
+
+  def process_batch_failures
+    return if !@batch["errors"] && all_stores_empty?
+
+    @batch["items"].each do |item|
+      process_document_failures(item)
+    end
+
+    AsyncIndexer.new(@indexer, "failures").enqueue_ids(@rerun_ids) unless @rerun_ids.empty?
+  end
+
+  def process_document_failures(item)
+    document = item[item.keys.first] # update/index/delete
+    document_stamp = { document["_id"].to_s => document["error"] }
+
+    first_store = get_store_as_json "first"
+    second_store = get_store_as_json "second"
+    permanent_store = get_store_as_json "permanent"
+
+    if !document["error"]
+      [first_store, second_store, permanent_store].each do |raw_store|
+         raw_store.select  { |doc| doc.keys.first == document["_id"] }.each do |doc|
+           raw_store.delete(doc)
+         end
+       end
+
+      set_store "first", first_store
+      set_store "second", second_store
+      set_store "permanent", permanent_store
+
+      return
+    end
+
+    return if permanent_store.include?(document_stamp)
+
+    if first_store.include?(document_stamp)
+      second_store << document_stamp
+      first_store.delete(document_stamp)
+      set_store "second", second_store
+      set_store "first", first_store
+      @rerun_ids << document["_id"]
```

Comment:
This looks like it would fail when used with the StatCounterIndexer. The IDs passed in as the arguments to the indexer aren't necessarily the same as the _id passed to ElasticSearch.

Reply:
Yeah, this needs to be addressed.

Reply:
I am looking into this now.

Reply:
Hi @tickinginstant -

I've looked through this, and I don't think I understand what you're getting at. Can you explain? It looks to me like anywhere the sweeper is called, it is specifically using the Elasticsearch "_id"s and never references anything else. I see that the StatCounterIndexer uses the StatCounter's work_id as it's Elasticsearch id - is this the basis of the problem that you're referring to?

Reply:
Yeah, that's the basis of the issue. The IndexSweeper reruns the indexer on this line:
```
AsyncIndexer.new(@indexer, "failures").enqueue_ids(@rerun_ids) unless @rerun_ids.empty?
```
When `@indexer` is the StatCounterIndexer, it'll pass in the `work_id`s for those StatCounters, because those are the `_id`s that will be returned from the ElasticSearch bulk update and stored in `@rerun_ids`. But the StatCounterIndexer is expecting to receive StatCounter IDs, not Work IDs (since it needs to integrate with the existing IndexQueue code). So in case of a failure, I believe the IndexSweeper will rerun the StatCounterIndexer with a completely different set of StatCounters than the failure occurred with (if those StatCounters even exist).

Reply:
Could this be solved by using the stat_counter id in the routing info for the StatCounterIndexer (the comment doesn't explain why the work_id is being used instead of the stat counter id)? Or, as in cases with BookmarkedSeries/BookmarkedExternalWorks/BookmarkedWorks, the id could be something like "#{stat_counter.id}-stat-counter", which would be read as the stat counter's id by the where clause on line 19 but would still differentiate the stat counters in the work index from the works.

But since I'm not entirely certain why it was done that way in the first place, I'm not sure what kind of problems that might cause.

Reply:
The StatCounterIndexer has to use the work_id in its routing info because it's not indexing StatCounters as a separate item. Instead, it's using the info from StatCounters to update works in the work index. So there's no way that it can be changed to use the StatCounter's ID.

The only way to use the StatCounter's ID is if you want to change the WorkIndexer so that it no longer includes stats, but instead generates a separate stats object alongside the work that gets joined with a has_parent/has_child query. But that's a more complex task, and I'm not sure if it's really a good idea, given that the ES docs kind of frown on has_parent/has_child queries except in specific circumstances:

> The join field shouldn’t be used like joins in a relation database. In Elasticsearch the key to good performance is to de-normalize your data into documents. Each join field, has_child or has_parent query adds a significant tax to your query performance.
>
> The only case where the join field makes sense is if your data contains a one-to-many relationship where one entity significantly outnumbers the other entity.

I could potentially see the additional cost being worth it if the need for joint queries was very rare, but I suspect that sorting works by kudos/comments/hits is one of the more common queries.

Reply:
Yes, the parent/child relationship is only barely worth it for bookmarks, and there's not much point to the stats indexing if you can't sort on the fields.

How much of a pain would it be to change the ids used in the StatCounterIndexer to work ids and to find the stat counter objects based on that? There's only one counter per work, so one should be as good as the other in theory.

Reply:
As I mentioned earlier, the problem with that is compatibility with IndexQueue and IndexSubqueue. If you want to have both indexers running simultaneously, you'd need to add some special-case code to IndexSubqueue to look up StatCounters based on work IDs, as well.

I feel like it'd be simpler to just have an extra class function in the Indexers, called something like `ids_from_elasticsearch`, which the IndexSweeper can pass the `@rerun_ids` through before passing them along to the AsyncIndexer. That would also have the bonus of being more flexible, so that you're not locked into always having to make the IDs match up.

Reply:
That sounds like a pretty good plan to me. @WendyBeth is this something you'd have time to look at this week, or should we make a separate issue for it?

Reply:
That sounds doable fairly quickly, and it's clear now why the StatCounterIndexer was made that way. Thanks for the information. I'll look into getting this cleared up tomorrow morning.